File: forge/features/observability/templates/enhanced_health/node/files/src/routes/health-deep.ts
import type { FastifyInstance } from "fastify";
import {
	checkKeycloak,
	checkRedis,
} from "../services/health-checks.service.js";

export async function healthDeepRoutes(app: FastifyInstance) {
	app.get("/deep", async (_req, reply) => {
		const [redis, keycloak] = await Promise.all([
			checkRedis(),
			checkKeycloak(),
		]);
		const overall = redis.status === "up" && keycloak.status === "up" ? "up" : "down";
		return reply.code(overall === "up" ? 200 : 503).send({
			status: overall,
			components: { redis, keycloak },
		});
	});
}

File: forge/features/observability/templates/enhanced_health/python/files/src/app/api/v1/endpoints/health_deep.py
"""Deep readiness endpoint that pings downstream services.

GET /api/v1/health/deep aggregates check_redis + check_keycloak alongside
the primary database check already performed by the base health router.
Returns 503 if any downstream component is DOWN so orchestrators (k8s,
Docker healthcheck) can react.
"""

from __future__ import annotations

from fastapi import APIRouter, Response, status

from app.core.health_checks import check_keycloak, check_redis

router = APIRouter()


@router.get("/deep")
async def deep_readiness(response: Response) -> dict:
    redis = await check_redis()
    keycloak = await check_keycloak()
    components = {
        "redis": redis.as_dict(),
        "keycloak": keycloak.as_dict(),
    }
    overall_up = redis.up and keycloak.up
    if not overall_up:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
    return {
        "status": "up" if overall_up else "down",
        "components": components,
    }
